# Караганда — квартиры
Знакомство с данными и их понимание (до EDA).

In [40]:
# Импорт библиотек
import pandas as pd

In [41]:
# Загрузка данных
df = pd.read_csv("karaganda_apartments.csv")

In [42]:
# Размер таблицы: (строки, столбцы)
df.shape

(200, 13)

In [43]:
# Первые 10 строк
df.head(10)

,id,title,rooms,area_m2,price_tenge,price_per_m2,address,city_date,url,balcony,parking,furnished,ceiling_height
0,693146384,1-комнатная квартира · 54.5 м²,1,54.5,30520000,560000,"Казыбек би р-н, Ержанова 63/2",Караганда 16 июн.,https://krisha.kz/a/show/693146384,NaN,NaN,NaN,3 м
1,693146389,3-комнатная квартира · 106.1 м²,3,106.1,59416000,560000,"Казыбек би р-н, Ержанова 63/2",Караганда 16 июн.,https://krisha.kz/a/show/693146389,NaN,NaN,NaN,3 м
2,693146382,2-комнатная квартира · 86.7 м²,2,86.7,48552000,560000,"Казыбек би р-н, Ержанова 63/2",Караганда 16 июн.,https://krisha.kz/a/show/693146382,NaN,NaN,NaN,3 м
3,1012319331,1-комнатная квартира · 30.5 м² · 5/5 этаж,1,30.5,15500000,508197,"Казыбек би р-н, мкр Новый Город, Ержанова 30",Караганда 16 июн.,https://krisha.kz/a/show/1012319331,NaN,NaN,NaN,NaN
4,693146383,2-комнатная квартира · 87.4 м²,2,87.4,48944000,560000,"Казыбек би р-н, Ержанова 63/2",Караганда 16 июн.,https://krisha.kz/a/show/693146383,NaN,NaN,NaN,3 м
5,693146387,3-комнатная квартира · 96.5 м²,3,96.5,54040000,560000,"Казыбек би р-н, Ержанова 63/2",Караганда 16 июн.,https://krisha.kz/a/show/693146387,NaN,NaN,NaN,3 м
6,1010848799,4-комнатная квартира · 100 м² · 6/17 этаж,4,100.0,55000000,550000,"Казыбек би р-н, мкр Городской Аэропорт, Букето...",Караганда 16 июн.,https://krisha.kz/a/show/1010848799,NaN,NaN,NaN,NaN
7,1011820130,2-комнатная квартира · 47 м² · 4/5 этаж,2,47.0,20900000,444681,"Казыбек би р-н, ул. Гоголя 49",Караганда 16 июн.,https://krisha.kz/a/show/1011820130,NaN,NaN,полностью,NaN
8,1012988997,3-комнатная квартира · 89 м² · 1/5 этаж,3,89.0,37500000,421348,"Казыбек би р-н, мкр Юго-Восток, Таттимбета 5/2",Караганда 16 июн.,https://krisha.kz/a/show/1012988997,NaN,NaN,NaN,3 м
9,1013145836,2-комнатная квартира · 63.5 м² · 1/19 этаж,2,63.5,51900000,817323,"Казыбек би р-н, мкр Новый Город, Мкр Байкена А...",Караганда 16 июн.,https://krisha.kz/a/show/1013145836,лоджия,NaN,NaN,NaN


In [44]:
# Последние 5 строк
df.tail()

,id,title,rooms,area_m2,price_tenge,price_per_m2,address,city_date,url,balcony,parking,furnished,ceiling_height
195,1012324369,2-комнатная квартира · 42.6 м² · 5/5 этаж,2,42.6,10700000,251174,Мкр. Восток-2 14,Караганда 16 июн.,https://krisha.kz/a/show/1012324369,NaN,NaN,NaN,NaN
196,1013094090,2-комнатная квартира · 47 м² · 5/5 этаж,2,47.0,9650000,205319,"Алихана Бокейханова р-н, Майкудук, Майкудук, 1...",Караганда 16 июн.,https://krisha.kz/a/show/1013094090,балкон,NaN,NaN,NaN
197,1011815796,1-комнатная квартира · 53 м² · 1/6 этаж,1,53.0,17800000,335849,"Алихана Бокейханова р-н, ​067-й учетный кварта...",Караганда 16 июн.,https://krisha.kz/a/show/1011815796,балкон,NaN,NaN,NaN
198,687864574,2-комнатная квартира · 48 м² · 4/5 этаж,2,48.0,13500000,281250,"Алихана Бокейханова р-н, 22 микрорайон 22",Караганда 16 июн.,https://krisha.kz/a/show/687864574,NaN,NaN,NaN,NaN
199,1011570466,3-комнатная квартира · 63.9 м² · 9/9 этаж,3,63.9,18000000,281690,"Алихана Бокейханова р-н, Майкудук 3 — ул. Стал...",Караганда 16 июн.,https://krisha.kz/a/show/1011570466,NaN,NaN,NaN,NaN


In [45]:
# Случайные 5 строк
df.sample(5)

,id,title,rooms,area_m2,price_tenge,price_per_m2,address,city_date,url,balcony,parking,furnished,ceiling_height
189,1013129109,3-комнатная квартира · 75 м² · 2/3 этаж,3,75.0,24500000,326667,"Казыбек би р-н, Жамбыла 7",Караганда 16 июн.,https://krisha.kz/a/show/1013129109,NaN,NaN,NaN,NaN
162,1013082349,2-комнатная квартира · 47 м² · 1/5 этаж,2,47.0,22000000,468085,"Казыбек би р-н, мкр Юго-Восток, 30й микрорайон...",Караганда 16 июн.,https://krisha.kz/a/show/1013082349,балкон,NaN,NaN,NaN
58,1011324462,3-комнатная квартира · 62.5 м² · 4/5 этаж,3,62.5,13000000,208000,"Алихана Бокейханова р-н, Майкудук, Майкудук, 1...",Караганда 16 июн.,https://krisha.kz/a/show/1011324462,NaN,NaN,NaN,2.5 м
84,1011128755,2-комнатная квартира · 59.6 м² · 1/9 этаж,2,59.6,27500000,461409,"Казыбек би р-н, мкр Юго-Восток, Шахтеров 31/2",Караганда 16 июн.,https://krisha.kz/a/show/1011128755,NaN,паркинг,NaN,NaN
8,1012988997,3-комнатная квартира · 89 м² · 1/5 этаж,3,89.0,37500000,421348,"Казыбек би р-н, мкр Юго-Восток, Таттимбета 5/2",Караганда 16 июн.,https://krisha.kz/a/show/1012988997,NaN,NaN,NaN,3 м


In [46]:
# Список столбцов
df.columns.tolist()

['id',
 'title',
 'rooms',
 'area_m2',
 'price_tenge',
 'price_per_m2',
 'address',
 'city_date',
 'url',
 'balcony',
 'parking',
 'furnished',
 'ceiling_height']

In [47]:
# Типы данных и общая информация
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              200 non-null    int64  
 1   title           200 non-null    object 
 2   rooms           200 non-null    int64  
 3   area_m2         200 non-null    float64
 4   price_tenge     200 non-null    int64  
 5   price_per_m2    200 non-null    int64  
 6   address         200 non-null    object 
 7   city_date       200 non-null    object 
 8   url             200 non-null    object 
 9   balcony         51 non-null     object 
 10  parking         6 non-null      object 
 11  furnished       2 non-null      object 
 12  ceiling_height  26 non-null     object 
dtypes: float64(1), int64(4), object(8)
memory usage: 20.4+ KB


In [48]:
# Типы данных по столбцам
df.dtypes

id                  int64
title              object
rooms               int64
area_m2           float64
price_tenge         int64
price_per_m2        int64
address            object
city_date          object
url                object
balcony            object
parking            object
furnished          object
ceiling_height     object
dtype: object

In [49]:
# Количество пропусков по столбцам
df.isnull().sum()

id                  0
title               0
rooms               0
area_m2             0
price_tenge         0
price_per_m2        0
address             0
city_date           0
url                 0
balcony           149
parking           194
furnished         198
ceiling_height    174
dtype: int64

In [50]:
# Доля пропусков в процентах
(df.isnull().mean() * 100).round(1)

id                 0.0
title              0.0
rooms              0.0
area_m2            0.0
price_tenge        0.0
price_per_m2       0.0
address            0.0
city_date          0.0
url                0.0
balcony           74.5
parking           97.0
furnished         99.0
ceiling_height    87.0
dtype: float64

In [51]:
# Количество полных дубликатов строк
df.duplicated().sum()

0

In [52]:
# Количество уникальных значений по столбцам
df.nunique()

id                200
title             192
rooms               6
area_m2           131
price_tenge       131
price_per_m2      175
address           181
city_date           1
url               200
balcony             4
parking             2
furnished           1
ceiling_height      5
dtype: int64

In [53]:
# Уникальные значения категориальных признаков
for col in ["rooms", "balcony", "parking", "furnished", "ceiling_height"]:
    print(col, "->", df[col].unique())

rooms -> [1 3 2 4 5 8]
balcony -> [nan 'лоджия' 'балкон' 'несколько балконов или лоджий' 'балкон и лоджия']
parking -> [nan 'паркинг' 'рядом охраняемая стоянка']
furnished -> [nan 'полностью']
ceiling_height -> ['3 м' nan '2.7 м' '2.6 м' '2.5 м' '3.2 м']


In [54]:
# Базовая статистика по числовым столбцам
df.describe()

,id,rooms,area_m2,price_tenge,price_per_m2
count,2.000000e+02,200.000000,200.000000,2.000000e+02,200.000000
mean,9.529733e+08,2.540000,69.997000,3.190694e+07,433865.015000
std,1.316707e+08,1.064927,35.118682,2.303919e+07,133096.268307
min,4.875918e+07,1.000000,14.000000,4.800000e+06,140625.000000
25%,1.002769e+09,2.000000,48.000000,1.697500e+07,345591.500000
50%,1.010918e+09,2.000000,63.950000,2.550000e+07,417618.000000
75%,1.012630e+09,3.000000,80.150000,3.800000e+07,512580.250000
max,1.013160e+09,8.000000,340.000000,1.900000e+08,850000.000000


In [55]:
# Статистика по всем столбцам, включая текстовые
df.describe(include="all")

,id,title,rooms,area_m2,price_tenge,price_per_m2,address,city_date,url,balcony,parking,furnished,ceiling_height
count,2.000000e+02,200,200.000000,200.000000,2.000000e+02,200.000000,200,200,200,51,6,2,26
unique,NaN,192,NaN,NaN,NaN,NaN,181,1,200,4,2,1,5
top,NaN,2-комнатная квартира · 48 м² · 3/5 этаж,NaN,NaN,NaN,NaN,"Казыбек би р-н, Ержанова 63/2",Караганда 16 июн.,https://krisha.kz/a/show/693146384,балкон,паркинг,полностью,3 м
freq,NaN,3,NaN,NaN,NaN,NaN,7,200,1,29,5,2,14
mean,9.529733e+08,NaN,2.540000,69.997000,3.190694e+07,433865.015000,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,1.316707e+08,NaN,1.064927,35.118682,2.303919e+07,133096.268307,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,4.875918e+07,NaN,1.000000,14.000000,4.800000e+06,140625.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,1.002769e+09,NaN,2.000000,48.000000,1.697500e+07,345591.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,1.010918e+09,NaN,2.000000,63.950000,2.550000e+07,417618.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,1.012630e+09,NaN,3.000000,80.150000,3.800000e+07,512580.250000,NaN,NaN,NaN,NaN,NaN,NaN,NaN


# Вопросы

In [56]:
# 1. Самая дорогая и самая дешёвая квартира
print("Максимум:", df["price_tenge"].max())
print("Минимум:", df["price_tenge"].min())
df.nlargest(5, "price_tenge")

Максимум: 190000000
Минимум: 4800000


,id,title,rooms,area_m2,price_tenge,price_per_m2,address,city_date,url,balcony,parking,furnished,ceiling_height
47,1013007081,8-комнатная квартира · 340 м²,8,340.0,190000000,558824,"Казыбек би р-н, мкр Михайловка , Мкр Михайловк...",Караганда 16 июн.,https://krisha.kz/a/show/1013007081,NaN,NaN,NaN,NaN
30,1001984483,5-комнатная квартира · 212.8 м²,5,212.8,138320000,650000,"Казыбек би р-н, мкр Новый Город, ​Проспект Нур...",Караганда 16 июн.,https://krisha.kz/a/show/1001984483,NaN,NaN,NaN,NaN
128,762112046,4-комнатная квартира · 142 м² · 3/5 этаж,4,142.0,99000000,697183,"Казыбек би р-н, Гоголя 56/1",Караганда 16 июн.,https://krisha.kz/a/show/762112046,NaN,NaN,NaN,NaN
79,1009437295,5-комнатная квартира · 158 м² · 9/10 этаж,5,158.0,95000000,601266,"Казыбек би р-н, Комиссарова 10/3",Караганда 16 июн.,https://krisha.kz/a/show/1009437295,NaN,NaN,NaN,3.2 м
35,1001816201,4-комнатная квартира · 140.9 м²,4,140.9,91585000,650000,"Казыбек би р-н, мкр Михайловка , Аманжолова 100",Караганда 16 июн.,https://krisha.kz/a/show/1001816201,NaN,NaN,NaN,NaN


In [57]:
# 2. Сколько квартир по количеству комнат
df["rooms"].value_counts()

rooms
2    70
3    64
1    32
4    29
5     4
8     1
Name: count, dtype: int64

In [58]:
# 3. Средняя цена за м² по числу комнат
df.groupby("rooms")["price_per_m2"].mean()

rooms
1    418121.531250
2    420603.114286
3    436166.156250
4    462050.586207
5    519492.750000
8    558824.000000
Name: price_per_m2, dtype: float64

In [59]:
# 4. Сколько квартир дороже 30 млн тенге (фильтр)
(df["price_tenge"] > 30_000_000).sum()

75

In [60]:
# 5. Средняя площадь у квартир с парковкой и без
df.groupby("parking")["area_m2"].mean()

parking
паркинг                     51.72
рядом охраняемая стоянка    14.00
Name: area_m2, dtype: float64

In [61]:
# 6. Топ-5 районов по числу объявлений
df["address"].str.extract(r"([\wА-Яа-я ]+? р-н)")[0].value_counts().head()

0
Казыбек би р-н             151
Алихана Бокейханова р-н     37
Name: count, dtype: int64

In [62]:
# 7. Сколько квартир меблированы
print(df["furnished"].value_counts())
print("Доля:", df["furnished"].notna().mean())

furnished
полностью    2
Name: count, dtype: int64
Доля: 0.01


In [63]:
# 8. Цена за м²: с балконом и без
df.groupby(df["balcony"].notna())["price_per_m2"].median()

balcony
False    432043.0
True     383784.0
Name: price_per_m2, dtype: float64

In [64]:
# 9. Самая большая квартира по площади
df.loc[df["area_m2"].idxmax(), ["area_m2", "price_tenge", "address"]]

area_m2                                                    340.0
price_tenge                                            190000000
address        Казыбек би р-н, мкр Михайловка , Мкр Михайловк...
Name: 47, dtype: object

In [65]:
# 10. Связь цены, площади и числа комнат (корреляция)
df[["area_m2", "price_tenge", "rooms"]].corr()

,area_m2,price_tenge,rooms
area_m2,1.000000,0.917586,0.807985
price_tenge,0.917586,1.000000,0.670598
rooms,0.807985,0.670598,1.000000


In [73]:
print("Максимум:", df["area_m2"].max())
print("Минимум:", df["area_m2"].min())
df.nlargest(5, "price_tenge")

Максимум: 340.0
Минимум: 14.0


,id,title,rooms,area_m2,price_tenge,price_per_m2,address,city_date,url,balcony,parking,furnished,ceiling_height,ppm2_calc,ppm2_diff,district,m2_per_room
47,1013007081,8-комнатная квартира · 340 м²,8,340.0,190000000,558824,"Казыбек би р-н, мкр Михайловка , Мкр Михайловк...",Караганда 16 июн.,https://krisha.kz/a/show/1013007081,NaN,NaN,NaN,NaN,558824,0,Казыбек би р-н,42.5
30,1001984483,5-комнатная квартира · 212.8 м²,5,212.8,138320000,650000,"Казыбек би р-н, мкр Новый Город, ​Проспект Нур...",Караганда 16 июн.,https://krisha.kz/a/show/1001984483,NaN,NaN,NaN,NaN,650000,0,Казыбек би р-н,42.6
128,762112046,4-комнатная квартира · 142 м² · 3/5 этаж,4,142.0,99000000,697183,"Казыбек би р-н, Гоголя 56/1",Караганда 16 июн.,https://krisha.kz/a/show/762112046,NaN,NaN,NaN,NaN,697183,0,Казыбек би р-н,35.5
79,1009437295,5-комнатная квартира · 158 м² · 9/10 этаж,5,158.0,95000000,601266,"Казыбек би р-н, Комиссарова 10/3",Караганда 16 июн.,https://krisha.kz/a/show/1009437295,NaN,NaN,NaN,3.2 м,601266,0,Казыбек би р-н,31.6
35,1001816201,4-комнатная квартира · 140.9 м²,4,140.9,91585000,650000,"Казыбек би р-н, мкр Михайловка , Аманжолова 100",Караганда 16 июн.,https://krisha.kz/a/show/1001816201,NaN,NaN,NaN,NaN,650000,0,Казыбек би р-н,35.2


**Что заметил:** есть квартира 14 м² и квартира 340 м² за 190 млн — крайности, которые тянут среднее.
**Как проверяю:** найду выбросы по цене формально — через межквартильный размах (правило `1.5·IQR`). Это те объекты, которые стоит проверять глазами и, возможно, исключать из моделей.

In [71]:
q1, q3 = df["price_tenge"].quantile([0.25, 0.75])
iqr = q3 - q1
low, high = q1 - 1.5 * iqr, q3 + 1.5 * iqr
print(f"Нормальный диапазон цены: {low:,.0f} — {high:,.0f} тг")

outliers = df[(df["price_tenge"] < low) | (df["price_tenge"] > high)]
print("Выбросов по цене:", len(outliers))
outliers[["rooms", "area_m2", "price_tenge", "price_per_m2", "address"]].sort_values("price_tenge", ascending=False)

Нормальный диапазон цены: -14,562,500 — 69,537,500 тг
Выбросов по цене: 10


,rooms,area_m2,price_tenge,price_per_m2,address
47,8,340.0,190000000,558824,"Казыбек би р-н, мкр Михайловка , Мкр Михайловк..."
30,5,212.8,138320000,650000,"Казыбек би р-н, мкр Новый Город, ​Проспект Нур..."
128,4,142.0,99000000,697183,"Казыбек би р-н, Гоголя 56/1"
79,5,158.0,95000000,601266,"Казыбек би р-н, Комиссарова 10/3"
35,4,140.9,91585000,650000,"Казыбек би р-н, мкр Михайловка , Аманжолова 100"
31,4,146.0,90520000,620000,"Казыбек би р-н, мкр Новый Город, ​Проспект Нур..."
22,3,131.2,87904000,670000,"Казыбек би р-н, мкр Новый Город, ​Улица Комисс..."
150,4,105.2,79500000,755703,"Казыбек би р-н, Таттимбета 3/14 — Гульдер 1"
125,4,105.0,78000000,742857,"Казыбек би р-н, мкр Юго-Восток, мкр Степной 3 ..."
112,4,143.0,70500000,493007,"Казыбек би р-н, мкр Юго-Восток, Просп. Республ..."


## Проверка адекватности: м² на комнату

**Зачем:** хороший способ поймать «битые» строки — площадь на одну комнату. Слишком мало (< 12 м²) или слишком много (> 60 м²) — повод присмотреться: ошибка в данных или нетипичный объект (студия / элитка).

In [74]:
df["m2_per_room"] = (df["area_m2"] / df["rooms"]).round(1)
print("Самые «тесные» (мало м² на комнату):")
print(df.nsmallest(5, "m2_per_room")[["title", "rooms", "area_m2", "m2_per_room", "price_tenge"]])


Самые «тесные» (мало м² на комнату):
                                         title  rooms  area_m2  m2_per_room  \
135    1-комнатная квартира · 14 м² · 2/5 этаж      1     14.0         14.0   
41     4-комнатная квартира · 62 м² · 2/5 этаж      4     62.0         15.5   
183  5-комнатная квартира · 86.9 м² · 4/5 этаж      5     86.9         17.4   
148    4-комнатная квартира · 72 м² · 3/9 этаж      4     72.0         18.0   
108  4-комнатная квартира · 72.3 м² · 3/3 этаж      4     72.3         18.1   

     price_tenge  
135      4800000  
41      22000000  
183     30000000  
148     28000000  
108     29000000  


In [75]:
print("\nСамые «просторные» на комнату:")
df.nlargest(5, "m2_per_room")[["title", "rooms", "area_m2", "m2_per_room", "price_tenge"]]


Самые «просторные» на комнату:


,title,rooms,area_m2,m2_per_room,price_tenge
0,1-комнатная квартира · 54.5 м²,1,54.5,54.5,30520000
197,1-комнатная квартира · 53 м² · 1/6 этаж,1,53.0,53.0,17800000
20,2-комнатная квартира · 103.2 м² · 2/9 этаж,2,103.2,51.6,67000000
131,1-комнатная квартира · 48 м² · 11/20 этаж,1,48.0,48.0,23000000
98,3-комнатная квартира · 138 м² · 9/18 этаж,3,138.0,46.0,52500000


## Что в итоге нашли
 
- **Район решает:** Казыбек би заметно дороже за м², чем Алихана Бокейханова 
- **Выбросы:** квартира 340 м² за 190 млн и микро-объект 14 м² — крайности, искажают среднее → для анализа лучше брать медиану 
